### PR TITLE
Program v3.1.0 (May Audit)

### PR DESCRIPTION
**Pull request by acamill**
### Fixed

- `Depository.getDeltaNeutralPositionNotionalSizeUI(mango: Mango)` now retrieves Mango account, PERP market & price concurrently.
- `Depository.getUnrealizedPnl(mango: Mango, options: ConfirmOptions)` now retrieves depository account & delta-neutral position size concurrently.
- `Depository.getFundingRate(mango: Mango)` now retrieves Mango Perp market & cache + bids & asks concurrently.
- `Depository.settleMangoDepositoryMangoAccountPnl(payer: Payer, mango: Mango)` now retrieves Mango account, cache, PERP market & quote root bank concurrently.

### Added

- Add `Client.createQuoteMintWithMangoDepositoryInstruction` method.
- Add `Client.createQuoteRedeemWithMangoDepositoryInstruction` method.
- Add `Client.createSetMangoDepositoryQuoteMintAndRedeemFeeInstruction` method.
- Add `Client.createSetMangoDepositoryQuoteMintAndRedeemSoftCapInstruction` method.
- Add `Client.createDisableDepositoryMintingInstruction` method.
- Add `Depository.getOffsetUnrealizedPnl(mango: Mango, options: ConfirmOptions)` method.
- Add `Depository.getCollateralPerpPriceNativeQuotePerNativeBase(mango: Mango): Promise<I80F48>` method.
- Add `Depository.getLimitPrice(slippage: I80F48, perpOrderTakerSide: 'short' | 'long', mango: Mango): Promise<I80F48> method.`
- Export new `findMultipleATAAddSync(wallet: PublicKey, mintAddresses: PublicKey[]): [PublicKey, number][]` utility function.

### Changed

- Update IDL to match UXD Program version 3.1.0 (audited).
- `Client.createRebalanceMangoDepositoryLiteInstruction` now returns `Promise<TransactionInstruction>` asynchronously to retrieve, compute & enforce the slippage limit price.
- `Client.createMintWithMangoDepositoryInstruction` now returns `Promise<TransactionInstruction>` asynchronously to retrive, compute & enforce the slippage limit price.
- `Client.createRedeemFromMangoDepositoryInstruction` now returns `Promise<TransactionInstruction>` asynchronously to retrieve, compute & enforce the slippage limit price.
- `Depository.getMinTradingSizeCollateralUI(mango: Mango)` now returns a `number` synchronously.
- `getBalance` utility function now throws an `Error` instead of a `string` when unable to retrieve token account balance.
- Upgrade `@project-serum/anchor` to `0.24.2` (from `0.22.0`).
- Upgrade `@solana/web3.js` to `1.42.0` (from `1.35.1`).
- Add `mangoDepositoriesQuoteRedeemableSoftCap!: BN; // u64` property to the Controller account.
- Add the following properties to the Depository account:
```
  netQuoteMinted!: BN; // i128  // The signed amount of quote-minted redeemable-tokens, negative means redeemed
  quoteMintAndRedeemFee!: number; // u8 // The fee applied to quote-mint/redeem
  totalQuoteMintAndRedeemFees!: BN; // u128 // The total amount of fee collected
  mintingDisabled!: boolean; // If the depository has disabled both collateral and quote minting
```
- All token account required by program instructions now need to be created by the caller. The program does not initialise accounts on behalf of the user anymore

### Removed

- Remove `findAddrSync` utility function export, use `PublicKey.findProgramAddressSync` instead.

### CODE REVIEW

**Comment by acamill** [comment]
@0xCryptoSheik  How to handle these issues with zero_one dep, can we ignore?
Screenshot 2022-05-06 at 10 58 53

File: src/01/depository.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+      case 'devnet':
+        zoProgramId = ZERO_ONE_DEVNET_PROGRAM_ID;
+        zoState = ZO_DEVNET_STATE_KEY;
+        break;
+      case 'mainnet-beta':
+        zoProgramId = ZERO_ONE_MAINNET_PROGRAM_ID;
+        zoState = ZO_MAINNET_STATE_KEY;
+        break;
+      default: {
+        throw 'error cluster';
```
Comment (0xCryptoSheik) [inline src/01/depository.ts]: `throw new Error('Invalid cluster')`

File: src/01/depository.ts [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    console.groupEnd();
+  }
+
+  public async getOnchainAccount(
+    connection: Connection,
+    options?: ConfirmOptions
+  ): Promise<ZoDepositoryAccount> {
+    const uxdProgram = new Program(IDL, this.uxdProgramId);
+
+    const ai = await connection.getAccountInfo(this.pda, options?.commitment);
```
Comment (0xCryptoSheik) [inline src/01/depository.ts]: `ai` ? Use a more descriptive variable name?

File: src/01/depository.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    return this.getPerpPriceNum(zo).n;
+  }
+
+  public getPerpPriceUI(zo: Zo): number {
+    return this.getPerpPriceNum(zo).number;
+  }
+
+  getPerpPriceNum(zo: Zo): Num {
+    const pmi = zo.state.getMarketIndexBySymbol(
+      this.collateralMintSymbol + '-PERP'
```
Comment (0xCryptoSheik) [inline src/01/depository.ts]: nit: \`${this.collateralMintSymbol}-PERP\`

File: src/01/depository.ts [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+      }
+    }
+  }
+
+  async getZoAccountCollateralOpenOrdersAccountKey(
+    zo: Zo,
+    controlKey: PublicKey
+  ): Promise<PublicKey> {
+    const dexMarket = await zo.state.getMarketBySymbol(
+      this.collateralMintSymbol + '-PERP'
```
Comment (0xCryptoSheik) [inline src/01/depository.ts]: same as above, use template string

File: src/01/depository.ts [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    )[0];
+  }
+
+  getZoTakerFee(): number {
+    return ZO_FUTURE_TAKER_FEE;
+  }
+
+  async minTradingSizeBase(zo: Zo): Promise<number> {
+    const dexMarket = await zo.state.getMarketBySymbol(
+      this.collateralMintSymbol + '-PERP'
```
Comment (0xCryptoSheik) [inline src/01/depository.ts]: same as above, use template string

File: src/01/depository.ts [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    );
+    return dexMarket.minOrderSize;
+  }
+
+  public async settleMarginAccountFunds(zo: Zo): Promise<string> {
+    let zoAccount = await zo.loadMarginAccount(this.zoAccountPda);
+    return zoAccount.settleFunds(this.collateralMintSymbol + '-PERP');
```
Comment (0xCryptoSheik) [inline src/01/depository.ts]: same as above, use template string

File: src/01/depository.ts [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  // async minTradingSizeQuote(zo: Zo): Promise<number> {
```
Comment (0xCryptoSheik) [inline src/01/depository.ts]: remove before merge if not included in this version

File: src/01/index.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+  State as zoState,
+  Margin as zoMargin,
+  Zo as ZoIdl,
+} from '@zero_one/client';
+import * as anchorZo from '@zero_one/client/node_modules/@project-serum/anchor';
+
+const invalidClusterError = 'Invalid cluster';
+
+// Helper to init and configure a Zo object.
+export async function createAndInitializeZo(
```
Comment (0xCryptoSheik) [inline src/01/index.ts]: remove `async`

File: src/01/index.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  Cluster as zoCluster,
+  createProvider as zoCreateProvider,
+  createProgram as zoCreateProgram,
+  State as zoState,
+  Margin as zoMargin,
+  Zo as ZoIdl,
+} from '@zero_one/client';
+import * as anchorZo from '@zero_one/client/node_modules/@project-serum/anchor';
+
+const invalidClusterError = 'Invalid cluster';
```
Comment (0xCryptoSheik) [inline src/01/index.ts]: inline if only used once, wrap in a `new Error(...)`

File: src/01/index.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  ZO_MAINNET_STATE_KEY,
+  Wallet as zoWallet,
+  Cluster as zoCluster,
+  createProvider as zoCreateProvider,
+  createProgram as zoCreateProgram,
+  State as zoState,
+  Margin as zoMargin,
+  Zo as ZoIdl,
+} from '@zero_one/client';
+import * as anchorZo from '@zero_one/client/node_modules/@project-serum/anchor';
```
Comment (0xCryptoSheik) [inline src/01/index.ts]: Uh oh!
Need to try align these.

File: src/01/index.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    state: zoState
+  ) {
+    this.program = program;
+    this.dexProgramId = zoDexProgram;
+    this.state = state;
+    // TO investigate, interval
+    state.startCacheRefreshCycle();
+  }
+
+  public async loadMarginAccount(owner: PublicKey): Promise<zoMargin> {
```
Comment (0xCryptoSheik) [inline src/01/index.ts]: remove `async`

File: src/01/index.ts [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    return await zoMargin.load(
```
Comment (0xCryptoSheik) [inline src/01/index.ts]: remove `await`

File: src/client.ts [hunk 13]
```diff
@@ -128,12 +126,15 @@ export class UXDClient {
...
     options: ConfirmOptions,
     payer?: PublicKey
-  ): TransactionInstruction {
-    const userCollateralATA = findATAAddrSync(
-      user,
-      depository.collateralMint
-    )[0];
-    const userQuoteATA = findATAAddrSync(user, depository.quoteMint)[0];
+  ): Promise<TransactionInstruction> {
+    const userCollateralATA = await utils.token.associatedAddress({
```
Comment (0xCryptoSheik) [inline src/client.ts]: wrap in a `Promise.all` since both can be done concurrently

File: src/client.ts [hunk 14]
```diff
@@ -160,13 +161,19 @@ export class UXDClient {
     const mangoPerpMarketConfig = mango.getPerpMarketConfig(
       depository.collateralMintSymbol
     );
-    const maxRebalancingAmountNative = new BN(
-      maxRebalancingAmount * 10 ** depository.quoteMintDecimals
-    );
+    const maxRebalancingAmountNative = uiToNative(maxRebalancingAmount, depository.quoteMintDecimals);
+    const perpSide = polarity == PnLPolarity.Positive ? 'short' : 'long'; //'sell' : 'buy';
+    const limit_price = (
+      await depository.getLimitPrice(
```
Comment (0xCryptoSheik) [inline src/client.ts:173]: Ideally this one too in the `Promise.all` of above but let's keep it simple for now

File: src/client.ts [hunk 15]
```diff
@@ -160,13 +161,19 @@ export class UXDClient {
     const mangoPerpMarketConfig = mango.getPerpMarketConfig(
       depository.collateralMintSymbol
     );
-    const maxRebalancingAmountNative = new BN(
-      maxRebalancingAmount * 10 ** depository.quoteMintDecimals
-    );
+    const maxRebalancingAmountNative = uiToNative(maxRebalancingAmount, depository.quoteMintDecimals);
+    const perpSide = polarity == PnLPolarity.Positive ? 'short' : 'long'; //'sell' : 'buy';
```
Comment (0xCryptoSheik) [inline src/client.ts]: `===`

File: src/client.ts [hunk 16]
```diff
@@ -332,33 +324,37 @@ export class UXDClient {
...
-    )[0];
-    const userRedeemableATA = findATAAddrSync(
-      user,
-      controller.redeemableMintPda
-    )[0];
-
-    const collateralAmountBN = new BN(
-      collateralAmount * 10 ** depository.collateralMintDecimals
-    );
+    const userCollateralATA = await utils.token.associatedAddress({
```
Comment (0xCryptoSheik) [inline src/client.ts]: wrap in a `Promise.all`

File: src/client.ts [hunk 17]
```diff
@@ -332,33 +324,37 @@ export class UXDClient {
...
+      mint: depository.collateralMint,
+      owner: user,
+    });
+    const userRedeemableATA = await utils.token.associatedAddress({
+      mint: controller.redeemableMintPda,
+      owner: user,
+    });
+    const collateralAmountBN = uiToNative(collateralAmount, depository.collateralMintDecimals);
+    const limit_price = (
+      await depository.getLimitPrice(
```
Comment (0xCryptoSheik) [inline src/client.ts:352]: Ideally wrap in the `Promise.all` of above, don't bother otherwise

File: src/client.ts [hunk 18]
```diff
@@ -332,33 +324,37 @@ export class UXDClient {
...
+        I80F48.fromNumber(slippage),
+        'short',
+        mango
+      )
+    ).toNumber();
+    console.log('LIMIT_PRICE for PERP ORDER : ', limit_price);
```
Comment (0xCryptoSheik) [inline src/client.ts]: No logs for merge

File: src/client.ts [hunk 19]
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
...
-      depository.collateralMint
-    )[0];
-    // Redeemable ATA
-    const userRedeemableATA = findATAAddrSync(
-      user,
-      controller.redeemableMintPda
-    )[0];
-    const redeemAmount = new BN(
-      amountRedeemable * 10 ** controller.redeemableMintDecimals
+    const userCollateralATA = await utils.token.associatedAddress({
```
Comment (0xCryptoSheik) [inline src/client.ts]: Promise.all

File: src/client.ts [hunk 20]
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
...
+      mint: depository.collateralMint,
+      owner: user,
+    });
+    const userRedeemableATA = await utils.token.associatedAddress({
+      mint: controller.redeemableMintPda,
+      owner: user,
+    });
+    const redeemAmountBN = uiToNative(amountRedeemable, controller.redeemableMintDecimals)
+    const limit_price = (
+      await depository.getLimitPrice(I80F48.fromNumber(slippage), 'long', mango)
```
Comment (0xCryptoSheik) [inline src/client.ts:427]: Promise.all if possible

File: src/client.ts [hunk 21]
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
...
+          tokenProgram: TOKEN_PROGRAM_ID,
+          mangoProgram: mango.programId,
+        },
+        options: options,
+      }
     );
-    return this.instruction.redeemFromMangoDepository(redeemAmount, slippage, {
+  }
+
+  public async createRegisterZoDepositoryInstruction(
```
Comment (0xCryptoSheik) [inline src/client.ts]: remove `async`

File: src/client.ts [hunk 22]
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
...
+    controller: Controller,
+    depository: ZoDepository,
+    zo: Zo,
+    control: PublicKey,
+    authority: PublicKey,
+    options: ConfirmOptions,
+    payer?: PublicKey
+  ): Promise<TransactionInstruction> {
+    const market = await zo.state.getMarketBySymbol(
+      depository.collateralMintSymbol + '-PERP'
```
Comment (0xCryptoSheik) [inline src/client.ts]: template string

File: src/client.ts [hunk 23]
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
...
+    );
+    const openOrdersPda =
+      await depository.getZoAccountCollateralOpenOrdersAccountKey(zo, control);
```
Comment (0xCryptoSheik) [inline src/client.ts]: in a `Promise.all` with the above

File: src/client.ts [hunk 24]
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
...
+    slippage: number,
+    controller: Controller,
+    depository: ZoDepository,
+    zo: Zo,
+    user: PublicKey,
+    options: ConfirmOptions,
+    payer?: PublicKey
+  ): Promise<TransactionInstruction> {
+    const marginAccount = await zo.loadMarginAccount(depository.pda);
+    const userCollateralATA = await utils.token.associatedAddress({
```
Comment (0xCryptoSheik) [inline src/client.ts]: `Promise.all` all the promises that can be done concurrently

File: src/client.ts [hunk 25]
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
...
+      mint: depository.collateralMint,
+      owner: user,
+    });
+    const userRedeemableATA = await utils.token.associatedAddress({
+      mint: controller.redeemableMintPda,
+      owner: user,
+    });
+    const limitPrice = depository.getLimitPrice(slippage, 'short', zo);
+    const market = await zo.state.getMarketBySymbol(
+      depository.collateralMintSymbol + '-PERP'
```
Comment (0xCryptoSheik) [inline src/client.ts]: template string

File: src/client.ts [hunk 26]
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
...
+    const maxQuoteQtyNativeBn = new BN(
+      Math.round(
+        limitPriceBn
+          .mul(maxBaseQtyBn)
+          .mul(market.decoded['quoteLotSize'])
+          .toNumber() * feeMultiplier
+      )
+    );
+    console.log('QUOTE LOT SIZE', market.decoded['quoteLotSize'].toNumber());
+    console.log('BASE LOT SIZE', market.decoded['baseLotSize'].toNumber());
```
Comment (0xCryptoSheik) [inline src/client.ts]: no logs for merge

File: src/client.ts [hunk 27]
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
...
+          zoAccount: depository.zoAccountPda,
+          zoState: zo.state.pubkey,
+          zoStateSigner: zo.state.signer,
+          zoCache: zo.state.cache.pubkey,
+          zoVault: zo.state.getVaultCollateralByMint(
+            depository.collateralMint
+          )[0],
+          zoControl: marginAccount.control.pubkey,
+          zoOpenOrders:
+            await depository.getZoAccountCollateralOpenOrdersAccountKey(
```
Comment (0xCryptoSheik) [inline src/client.ts]: avoid putting `await` of async operations in long-ass object declarations, prefer `await` in a variable and then use that later on

File: src/client.ts [hunk 28]
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
...
+    slippage: number,
+    controller: Controller,
+    depository: ZoDepository,
+    zo: Zo,
+    user: PublicKey,
+    options: ConfirmOptions,
+    payer?: PublicKey
+  ): Promise<TransactionInstruction> {
+    const marginAccount = await zo.loadMarginAccount(depository.pda);
+    const userCollateralATA = await utils.token.associatedAddress({
```
Comment (0xCryptoSheik) [inline src/client.ts]: `Promise.all`

File: src/client.ts [hunk 29]
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
...
+      owner: user,
+    });
+    const userRedeemableATA = await utils.token.associatedAddress({
+      mint: controller.redeemableMintPda,
+      owner: user,
+    });
+    const perpPrice = depository.getPerpPriceUI(zo);
+    const limitPrice = depository.getLimitPrice(slippage, 'long', zo);
+    const market = await zo.state.getMarketBySymbol(
+      depository.collateralMintSymbol + '-PERP'
```
Comment (0xCryptoSheik) [inline src/client.ts]: template string

File: src/client.ts [hunk 30]
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
...
+    const feeMultiplier = 1 - ZO_FUTURE_TAKER_FEE;
+    // We reduce the max quote by the max fees amount so that they can be collected on top of the redeemed amount of token
+    // (See program redeem instruction comment for more infos)
+    const maxQuoteQtyLotPurchaseBn = market.quoteSizeNumberToLots(
+      redeemableAmount * feeMultiplier
+    );
+    const maxBaseQtyBn = new BN(
+      Math.round(maxQuoteQtyLotPurchaseBn.div(perpPriceBn).toNumber())
+    );
+    console.log('Limit price', limitPriceBn.toNumber());
```
Comment (0xCryptoSheik) [inline src/client.ts]: no log for merge

File: src/client.ts [hunk 31]
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
...
+          zoAccount: depository.zoAccountPda,
+          zoState: zo.state.pubkey,
+          zoStateSigner: zo.state.signer,
+          zoCache: zo.state.cache.pubkey,
+          zoVault: zo.state.getVaultCollateralByMint(
+            depository.collateralMint
+          )[0],
+          zoControl: marginAccount.control.pubkey,
+          zoOpenOrders:
+            await depository.getZoAccountCollateralOpenOrdersAccountKey(
```
Comment (0xCryptoSheik) [inline src/client.ts]: outside of the object declaration please

File: src/client.ts [hunk 32]
```diff
@@ -405,51 +398,353 @@ export class UXDClient {
...
+  public async createWithdrawInsuranceFromZoDepositoryInstruction(
+    insuranceWithdrawnAmount: number,
+    controller: Controller,
+    depository: ZoDepository,
+    zo: Zo,
+    authority: PublicKey,
+    options: ConfirmOptions
+  ): Promise<TransactionInstruction> {
+    const marginAccount = await zo.loadMarginAccount(depository.pda);
+    const authorityQuoteATA = await utils.token.associatedAddress({
```
Comment (0xCryptoSheik) [inline src/client.ts]: Promise.all

File: src/mango/depository.ts [hunk 33]
```diff
@@ -263,6 +280,62 @@ export class MangoDepository {
     return pm.getCurrentFundingRate(mango.group, mc, pmi, bids, asks);
   }
 
+  // Return the depository lifetime funding PnL
+  public async getFundingPnl(): Promise<number> {
```
Comment (0xCryptoSheik) [inline src/mango/depository.ts]: ok lol nice

File: src/mango/depository.ts [hunk 34]
```diff
@@ -263,6 +280,62 @@ export class MangoDepository {
...
+    quote: { deposit: number; borrow: number };
+    collateral: { deposit: number; borrow: number };
+  }> {
+    let usdcBorrowInterest: number = 0;
+    let usdcDepositInterest: number = 0;
+    let collateralBorrowInterest: number = 0;
+    let collateralDepositInterest: number = 0;
+    let mangoCache = await mango.getCache();
+
+    var response = await fetch(
```
Comment (0xCryptoSheik) [inline src/mango/depository.ts]: Promise.all with the previous

**Comment by 0xCryptoSheik** [comment]
> BREAKING: Some transaction creation instructions now return a promise instead - I think this is bad because it's now blocking? What is the best way, inject the needed accounts?

It's fine, though it'd be good to check a little deeper as to why this changed & other potential impact

**Comment by 0xCryptoSheik** [comment]
> BREAKING: Upgrade anchor to 0.24.2 (from 0.22.0)

Ideally when we start using this, have everything aligned to this version of anchor (staking, FE, client, mango client, zo client, program).

**Comment by 0xCryptoSheik** [comment]
> @0xCryptoSheik How to handle these issues with zero_one dep, can we ignore? Screenshot 2022-05-06 at 10 58 53

For now yes ignore, but ideally need to find the type definitions for bn.js, I'll look into it soon

**Review by 0xCryptoSheik — CHANGES_REQUESTED** [review]
See comments

**Comment by 0xCryptoSheik** [comment]
I'm surprised we don't already have the issue with BN, anchor must have provided their own declarations.

**Comment by 0xCryptoSheik** [comment]
Aight typechecking errors are fixed/suppressed, will make another pass before merge, don't hesitate to run `npm run validate` from time to time

**Comment by 0xCryptoSheik** [comment]
> Add methods to retrieve Pnl for borrow/deposit/funding from Mango Stats in MangoDepository

@acamill can we add this as part of its own PR?

Sounds unrelated to the program update & 01 support & has a few considerations of its own.

File: src/01/depository.ts (hunk 7 above)
Comment (0xCryptoSheik) [reply src/01/depository.ts]: @acamill keep & finish or remove?

File: src/01/depository.ts [hunk 35]
```diff
@@ -0,0 +1,183 @@
...
+      }
+    }
+    this.collateralMint = mint;
+    this.collateralMintSymbol = mintName;
+    this.collateralMintDecimals = mintDecimals;
+    this.quoteMint = quoteMint;
+    this.quoteMintSymbol = quoteMintName;
+    this.quoteMintDecimals = quoteMintDecimals;
+    this.pda = findAddrSync(
+      [Buffer.from('ZODEPOSITORY'), mint.toBuffer()],
```
Comment (0xCryptoSheik) [inline src/01/depository.ts]: `Buffer` is a Node.js API, it's not available in the browser, which means consumers who'll want to use `uxd-client` for an app, will need to tweak their bundler configuration to include a browser-compatible `Buffer` module & make it available for global use.

I didn't see we did that as well in other places, it's fine for now but better not to do this & include a buffer module dependency instead at some point.

**Comment by 0xCryptoSheik** [comment]
> > Add methods to retrieve Pnl for borrow/deposit/funding from Mango Stats in MangoDepository
> 
> @acamill can we add this as part of its own PR?
> 
> Sounds unrelated to the program update & 01 support & has a few considerations of its own.

Confirmed this, extracting & opening a new MR

File: src/client.ts (hunk 14 above)
Comment (0xCryptoSheik) [reply src/client.ts:173]: also, camelcase

File: src/01/depository.ts (hunk 35 above)
Comment (0xCryptoSheik) [reply src/01/depository.ts]: Leaving as-is for now, may fix in another PR

**Comment by 0xCryptoSheik** [comment]
removing everything related to 01 too

File: src/client.ts [hunk 36]
```diff
@@ -17,6 +18,7 @@ import { findATAAddrSync } from './utils';
 import NamespaceFactory from './namespace';
 import { IDL as UXD_IDL } from './idl';
 import { PnLPolarity } from './interfaces';
+import { I80F48 } from './index';
```
Comment (0xCryptoSheik) [inline src/client.ts]: rather from Mango

File: src/client.ts [hunk 37]
```diff
@@ -128,12 +132,22 @@ export class UXDClient {
...
-      user,
-      depository.collateralMint
-    )[0];
-    const userQuoteATA = findATAAddrSync(user, depository.quoteMint)[0];
+  ): Promise<TransactionInstruction> {
+    const perpSide = polarity === PnLPolarity.Positive ? 'short' : 'long'; //'sell' : 'buy';
+
+    const [userCollateralATA, userQuoteATA, limitPriceI80F48] =
+      await Promise.all([
+        utils.token.associatedAddress({
```
Comment (0xCryptoSheik) [inline src/client.ts]: @acamill so basically `findATAAddrSync` is deprecated in favor of ` utils.token.associatedAddress` ?

File: src/client.ts [hunk 38]
```diff
@@ -452,4 +526,102 @@ export class UXDClient {
...
+    const mangoDepositedVaultAccount = mango.getVaultFor(depositedTokenIndex);
+    const mangoPerpMarketConfig = mango.getPerpMarketConfig(
+      depository.collateralMintSymbol
+    );
+    const userQuoteATA = findATAAddrSync(user, depository.quoteMint)[0];
+    const userRedeemableATA = findATAAddrSync(
+      user,
+      controller.redeemableMintPda
+    )[0];
+    const redeemableAmountNativeBN = new BN(
```
Comment (0xCryptoSheik) [inline src/client.ts]: uiToNative ?

File: src/client.ts [hunk 39]
```diff
@@ -405,42 +428,93 @@ export class UXDClient {
...
+    const mangoRootBankAccount = mango.getRootBankForToken(depositedTokenIndex);
+    const mangoNodeBankAccount = mango.getNodeBankFor(
+      depositedTokenIndex,
+      depository.quoteMint
+    );
+    const mangoDepositedVaultAccount = mango.getVaultFor(depositedTokenIndex);
+    const mangoPerpMarketConfig = mango.getPerpMarketConfig(
+      depository.collateralMintSymbol
+    );
+    const userQuoteATA = findATAAddrSync(user, depository.quoteMint)[0];
```
Comment (0xCryptoSheik) [inline src/client.ts]: `findATAAddrSync` or `utils.token.associatedAddress` ?

There can be only one.

File: src/client.ts [hunk 40]
```diff
@@ -405,42 +428,93 @@ export class UXDClient {
...
-    const redeemAmount = new BN(
-      amountRedeemable * 10 ** controller.redeemableMintDecimals
+    const quoteAmountNativeBN = new BN(
+      quoteAmount * 10 ** depository.collateralMintDecimals
     );
-    return this.instruction.redeemFromMangoDepository(redeemAmount, slippage, {
+    return this.instruction.quoteMintWithMangoDepository(quoteAmountNativeBN, {
       accounts: {
-        user,
+        user: user,
```
Comment (0xCryptoSheik) [inline src/client.ts:497]: just `user`, is good

File: src/client.ts [hunk 41]
```diff
@@ -405,42 +428,93 @@ export class UXDClient {
...
+      depository.collateralMintSymbol
+    );
+    const userQuoteATA = findATAAddrSync(user, depository.quoteMint)[0];
     const userRedeemableATA = findATAAddrSync(
       user,
       controller.redeemableMintPda
     )[0];
-    const redeemAmount = new BN(
-      amountRedeemable * 10 ** controller.redeemableMintDecimals
+    const quoteAmountNativeBN = new BN(
```
Comment (0xCryptoSheik) [inline src/client.ts]: `uiToNative` ?

File: src/client.ts [hunk 42]
```diff
@@ -452,4 +526,102 @@ export class UXDClient {
...
+    const mangoRootBankAccount = mango.getRootBankForToken(depositedTokenIndex);
+    const mangoNodeBankAccount = mango.getNodeBankFor(
+      depositedTokenIndex,
+      depository.quoteMint
+    );
+    const mangoDepositedVaultAccount = mango.getVaultFor(depositedTokenIndex);
+    const mangoPerpMarketConfig = mango.getPerpMarketConfig(
+      depository.collateralMintSymbol
+    );
+    const userQuoteATA = findATAAddrSync(user, depository.quoteMint)[0];
```
Comment (0xCryptoSheik) [inline src/client.ts]: findATAAddrSync vs utils.token.blah

File: src/interfaces.ts [hunk 43]
```diff
@@ -34,7 +34,24 @@ export class MangoDepositoryAccountV1 {
   totalAmountPaidTakerFee!: BN; // u128
 }
 
-// V2
+// V2 with ZO no USDC
```
Comment (0xCryptoSheik) [inline src/interfaces.ts]: Can we haz better comments please?

File: src/mango/depository.ts [hunk 44]
```diff
@@ -290,4 +292,35 @@ export class MangoDepository {
...
+    mango: Mango
+  ): Promise<I80F48> {
+    const mangoCache = await mango.getCache();
+    const pmc = this.getPerpMarketConfig(mango); // perpMarketConfig
+    const pmi = pmc.marketIndex; // perpMarketIndex
+    return mango.group.getPriceNative(pmi, mangoCache);
+  }
+
+  // The side of the taker (the user)
+  async getLimitPrice(
```
Comment (0xCryptoSheik) [inline src/mango/depository.ts:356]: assuming these 2 methods are now required for 3.1.0

File: src/client.ts (hunk 37 above)
Comment (acamill) [reply src/client.ts]: I think it's the new interface in spl token 0.2.0 but I was accessing it through 01.. cheater.

I think we can roll back to the 0.1.8 way, I did in the program.

If I'm not mistaken.

File: src/client.ts (hunk 38 above)
Comment (acamill) [reply src/client.ts]: Yes preferable anywhere this is used

File: src/client.ts (hunk 39 above)
Comment (acamill) [reply src/client.ts]: FindAtaAddrSync let's not change what doesn't need to be change

File: src/mango/depository.ts (hunk 44 above)
Comment (acamill) [reply src/mango/depository.ts:356]: Yes!

File: src/client.ts (hunk 37 above)
Comment (0xCryptoSheik) [reply src/client.ts]: Mhh, you're accessing it from `@project-serum/anchor`.

I'll take a closer look at the differences between the two & migrate to the best.

File: src/client.ts (hunk 39 above)
Comment (0xCryptoSheik) [reply src/client.ts]: https://github.com/UXDProtocol/uxd-client/pull/3#discussion_r875697863 if that's ok

File: src/client.ts (hunk 37 above)
Comment (0xCryptoSheik) [reply src/client.ts]: PSA: please do not use `utils.token.associatedAddress` from anchor, or any async `findProgramAddress` .
It was originally async in `@solana/web3.js`, then they changed it to be synchronous, kept the async version for backwards compatibility.
TLDR: use `PublicKey.findProgramAddressSync`
I will change it in this branch to use only the sync version, and remove any use of `utils.token.associatedAddress` from anchor

File: src/client.ts (hunk 39 above)
Comment (acamill) [reply src/client.ts]: Yes! I did change some stuff that wasn't needed, I mean that was the meaning of my "let's not change..." above 😁 
Thanks for the keen eye

**Comment by acamill** [comment]
@0xCryptoSheik Thanks for handling this, pushed the last tiny changes

**Comment by 0xCryptoSheik** [comment]
poke @acamill 

```
src/mango/depository.ts
[error] src/mango/depository.ts: SyntaxError: Declaration or statement expected. (344:3)
[error]   342 |   // `collateralQuantity` : In Collateral UI units
[error]   343 |   //  Output -> breakdown of estimated costs
[error] > 344 |   public getMintingEstimates(
[error]       |   ^
[error]   345 |     collateralQuantity: number,
[error]   346 |     perpPrice: number,
[error]   347 |     mintingPriceImpact: number,
```

**Comment by 0xCryptoSheik** [comment]
poke @acamill 

```
 FAIL  src/__tests__/mango-price-impact.spec.ts
  Price Impact / Mint/Redeem estimations tests
    ✕ Minting (138 ms)
    ✕ Redeeming (73 ms)

  ● Price Impact / Mint/Redeem estimations tests › Minting

    Price impact couldn't be determined.

      424 |     const priceImpact = sideBook.getImpactPriceUi(nativeQuantity);
      425 |     if (typeof priceImpact === 'undefined') {
    > 426 |       throw new Error("Price impact couldn't be determined.");
          |             ^
      427 |     }
      428 |     return priceImpact;
      429 |   }

      at MangoDepository.getMangoPerpOrderBookPriceImpact (src/mango/depository.ts:426:13)
      at Object.<anonymous> (src/__tests__/mango-price-impact.spec.ts:29:32)
```
